addons/web_editor/static/src/js/editor/odoo-editor/src/commands/commands.js

```diff
@@ -182,21 +182,30 @@ export const editorCommands = {
         // In case the html inserted is all contained in a single root <p> or <li>
         // tag, we take the all content of the <p> or <li> and avoid inserting the
         // <p> or <li>. The same is true for a <pre> inside a <pre>.
+        const isBlockElement = (node) => isBlock(node) && !["TABLE", "UL", "OL", "DIV"].includes(node.nodeName) && startNode.textContent !== "";
+
         if (container.childElementCount === 1 && (
             container.firstChild.nodeName === 'P' ||
             container.firstChild.nodeName === 'LI' ||
-            container.firstChild.nodeName === 'PRE' && closestElement(startNode, 'pre')
+            container.firstChild.nodeName === 'PRE' && closestElement(startNode, 'pre') ||
+            isBlockElement(container.firstChild)
         )) {
             const p = container.firstElementChild;
             container.replaceChildren(...p.childNodes);
         } else if (container.childElementCount > 1) {
             // Grab the content of the first child block and isolate it.
-            if (isBlock(container.firstChild) && !['TABLE', 'UL', 'OL'].includes(container.firstChild.nodeName)) {
+            if (
+                isBlockElement(container.firstChild) ||
+                container.firstChild.nodeName === "P"
```

Comment:
I don't get this. We'll be unwrapping the first child's contents if it's a block and not a table/list/div and has text, or if it's a paragraph? But paragraphs are already blocks and not table/list/div so basically all this seems to change is that we also enter the condition if the first child is an empty paragraph. How does that relate with the heading issue?

Reply:
> We'll be unwrapping the first child's contents if it's a block and not a table/list/div and has text, or if it's a paragraph?

> **_and has text_**

Noo, it says startNode.textContent !== '' not container.firstChild.textContent !== '' not the container.firstChild but the node on which the text is pasted is has text... That was the very reason I want to move that condition out of `isBlockElement`

> How does that relate with the heading issue?

This might have been done to maintain the test case:
contentBefore: `<p>[]<br></p>`
pasteHTML: `<p>abcd</p><p>efgh</p>`
contentAfter: `<p>abcd</p><p>efgh[]</p>`

Without the code suggested by the commit, an issue would arise, where an empty `<p>` tag would be kept below the pasted content, resulting in: `<p>abcd</p><p>efgh</p><p>[]<br></p>`

To maintain the intended behavior, the commit introduced a condition specifically for `<p>` tags by checking container.firstChild.nodeName === "P" and a similar check for the lastChild. However, the commit should have focused on removing empty `<p>` tag below it. Rather it kept `<p>` tag as an exception.

I have **fixed** it.